### DISCUSSION TITLE
Spanish punctuation symbols

### DISCUSSION THREAD

**Discussion by Lobo-Feroz** [binarybottle/engram-es#7] [Ideas]
Here's a comprehensive list of spanish punctuation symbols. Source: https://www.uv.es/normas/2011/materiales/Signos_2011.pdf (University of Valencia)

They are listed alphabetically as in the doc, will cover their frequency separately.

1. <> - Antilambda - Diple. Not very used nowadays except as "less/more than". https://en.wikipedia.org/wiki/Diple_(textual_symbol)
2. ' - Apostrophe. Not very used in spanish language. https://en.wikipedia.org/wiki/Apostrophe
3. \* \- Asterisk. Used to call attention, e.g. to a footnote. https://en.wikipedia.org/wiki/Asterisk
4. | - Bar. Not used too much in the language itself, but used in computers. https://en.wikipedia.org/wiki/Vertical_bar
5. ¶ - Calderón - Pilcrow. Not used in the modern spanish language. No need to have it in the layout, or maybe with some non-trivial key combo. https://en.wikipedia.org/wiki/Pilcrow
6. , - Comma. Ubiquitous, as in english.
7. («  ») (“ ”) (‘’) - Comillas - Quotation Marks. There are three types of them: spanish («  »), english (“ ”) and simple (‘’)
Nowadays nobody would tell, but the spanish quotations are technically the preferred way to quote. They've been displaced by the english because they're absent from qwerty-es. The simple quotes are different for opening and for closing, but in this forum they are both shown as apostrophes. You can check their actual shape here: https://en.wikipedia.org/wiki/Quotation_mark
In english, the spanish quotations are called "Guillemets": https://en.wikipedia.org/wiki/Guillemet
8. ( [ ] ) - Corchete - Square Bracket. Used for additional information, but parenthesis are preferred. https://en.wikipedia.org/wiki/Bracket#Square_brackets
9. ¨ - Diaeresis. Makes a silent u vowel have sound. https://en.wikipedia.org/wiki/Diaeresis_(diacritic)
In qwerty-es diaeresis can be aplied to all vowels ÄËÏÖÜ though it only makes sense in the U.
10. : - Dos Puntos - Colon -  used to show an explanation or an enumeration: just like in english.
11. (→←↑↓↔↕) - Arrow - it's actually considered a symbol in spanish :) I think we can keep switching - - > to →
12. ( - ) - Guión - Hyphen. Quite the same as in english. In spanish you can also use it to break a word at the end of a line, but this has been also made obsolete since word processors can justify text.
13. _ - Guión bajo - Underscore. Same as in english.
14. { } - Llave - Braces. Not very used in the language, but they are used in computer code.
15. ( ) - Parenthesis - Same as in english.
16. § - Paragraph. Not very used in the language nor in computers. Skippable or non-trivial key combo.
17. ( . ) - Punto - Dot. Same as in english.
18. ... - Puntos suspensivos - Ellipsis. Same as in english, to "suspend" the sentence: one, two, three... No need to have as a separate symbol as all word processors turn three dots into an ellipsis.
19. ; - Punto y coma - Semicolon. Same as in english, mainly to separate elements in an enumeration, and in code.
20. — - Raya - Dash. A longer hyphen for lists, dialogues, explanations... https://en.wikipedia.org/wiki/Dash
21. (¿?) (¡!) - Interrogations and exclamations. The difference is that we also have an opening sign, like with parenthesis. ¿What? ¡Yeah!
22. ´ - Tilde. Goes on top of a vowel and denotes the accented syllabe of the word. áéíóú ÁÉÍÓÚ. We're discussing these in a separate thread.

**Comment by binarybottle** [comment]
Please tell me what you think about my current plan is to keep all of the punctuation where it is in the English Engram layout, but to include the following:

¿ ¡
Simultaneously press the ? or ! key and the right-alt key.

« »
Simultaneously press either the ' or " key (or perhaps the < or > key) and the right-alt key for the left and right Guillemets.

**Reply by Lobo-Feroz** [reply]
I don't know if keeping exactly the same engram layout for punctuation is the best idea. Not because it isn't a good layout, but because spanish punctuation is different. 

The main difference being that the apostrophe is very, very rarely used in spanish. So that could free one key, which could go to either the tilde ( ´ ) or to separate ( ? ) and ( ! )

~~As for the opening signs, my first idea would be to keep them consistent, i.e. using shift plus the key, for the opening sign. My rationale for this is that we're used to capitalize the first character of a sentence, so we're kinda trained to use shift to start a sentence. But this is just an idea, of course.~~

~~? + shift: ¿~~
~~! + shift: ¡~~
~~\> + shift: <~~

Edit: on second thoughts, I think it's more cohesive to keep opening and closing signs on left-right keys like engram-en already does for diples and parenthesis. See next comment.

Also the Guillemets being the AltGr (or tilded) version of the Diple seems like the best solution to me as well! 

\> + AlGr (or tilde): »
\< + AlGr (or tilde): «

I think at this point the actual symbol frequencies would be useful. I'll try and see if I can have a shot at them.

**Reply by Lobo-Feroz** [reply]
Now that I think of it, another possibility for opening & closing exclamations and interrogations would be to separate them spatially, like you have on the current engram-en 2.0 with the parenthesis, diples, brackets and curly brackets. Specially the parenthesis in the middle columns upper row.

We could have the opening signs in the middle right column, lower row, just where the hyphen is. ( ¿ ) for unshifted and ( ! ) for shifted. 

This way every punctuation sign in the keyboard would have its opening sign at the left, its closing sign at the right.

This would mean we'd have to move the hyphen and underscore somewhere else, but we will be freeing several keys anyway since we won't use Ñ, Ç or ( ` ). Also the spanish diple key (left of Z) will be freed, since they're on the 6-7 keys. The hyphen-underscore (or any other sign, depending on frequency calculations) could go there.

It's an accessible key just moving the pinky a little bit downwards. The problem though: it's not ANSI, only ISO.

**Reply by iandoug** [reply]
What about the 'single' versions of  » and « ?

**Reply by Lobo-Feroz** [reply]
Hey Ian, nice to see you around here!

Do you mean the Diples? https://en.wikipedia.org/wiki/Diple_(textual_symbol)
If so, they are the shift versions of the keys 5 and 6 (as per the layout at https://engram.dev/)

If you mean some other symbol, could you please specify?

**Comment by iandoug** [comment]
I mean ‹› 

https://en.wikipedia.org/wiki/Guillemet

**Reply by Lobo-Feroz** [reply]
I don't think single guillemets are used in the spanish language.

The pdf I linked in the first post ([this one]( https://www.uv.es/normas/2011/materiales/Signos_2011.pdf )), from the University of Valencia, claims to be an exhaustive list of signs. It does mention diples but not single guillemets.

I've searched a bit in the internet and I can't find single guillemets in any other list of spanish symbols.

And, though this is anecdotal, as a spaniard, I don't recall having ever seen them in written spanish text. 

Wikipedia ([quotation marks](https://en.wikipedia.org/wiki/Quotation_mark#History)) says:

> The reemergence of single quotation marks around 1800 came about as a means of indicating a secondary level of quotation.[citation needed] In some languages using the angular quotation marks, the usage of the single guillemet, ‹…›, became obsolete, being replaced by double curved ones: “…”, though the single ones still survive, for instance, in Switzerland. In Eastern Europe, the curved quotation marks, „…“, are used as a secondary level when the angular marks, «…» are used as a primary level. 

In spanish is as the last sentence says, « » guillemets for primary level quotes, "" for secondary, '' for tertiary.

So my conclusion at this point would be that it doesn't seem necessary to include the single guillemets in the engram-es keyboard.

**Comment by iandoug** [comment]
From Wiki page:
Spanish (uncommon in daily usage, but commonly used in publishing) 
(although this is not specific to the single version)

How about Galician or Catalan?

That quick corpus (one file) analysis I did has

“	313	0.00034
»	309	0.00034
«	301	0.00033
”	300	0.00033
but no single versions. Let me check some books, the web or Wiki does not have a lot of dialogue ...

Cheers, Ian

**Reply by Lobo-Feroz** [reply]
Yes, I think the "uncommon in daily usage, but commonly used in publishing" refers to the spanish quotation marks or double guillemets.

They are used in publications because it's the canonical way to quote. But they are not used in daily usage texts because... guess what? they didn't include them in the qwerty-es keyboard ! So nowadays everybody here uses the "" for quotations.

Qwerty is truly a marvel of bad design!

**Reply by iandoug** [reply]
> Qwerty is truly a marvel of bad design!

Probably back in the 70s/80s, the likes of IBM thought, "What, make a different keyboard for our 5 customers in Spain / France / Germany?" Nah, they can use the same keyboards as us ...

**Reply by Lobo-Feroz** [reply]
Well yeah, not meant to disparage. They worked with whatever resources they had at the time.

But now we have the resources, specially Arno's mind :), to create something way better...

**Comment by iandoug** [comment]
Okay checked several book on Gutenberg, did not find single version used.
Thanks.

**Comment by binarybottle** [comment]
I was thinking about a shifted version of the ? and ! producing the upside-down versions of both, but you've got me thinking about left-right pairs of keys just like parentheses. I'll get on this as soon as I'm finished re-running my script to optimize layouts with fewer restrictions than I used for my first round.

**Comment by binarybottle** [comment]
What do you think about the following strategy,  @NickG13, @Lobo-Feroz, and @sunaku?  
All diacriticals and some symbols are accessed by a special key (denoted by a star):

star + A,E,I,O,U = ÁÉÍÓÚ
star + N = Ñ
star + C = Ç
star + : = Ü
star + «» = "'
star + ,; = -_
star + ?! = ¿¡
star + ) = `
star + € = $

![engram-es-diacriticals](https://user-images.githubusercontent.com/236532/127756450-bb6f5d7f-552a-4cd6-b973-66f7cc8d01ef.jpg)

**Reply by iandoug** [reply]
Arno, think you should use ISO not ANSI?
so "A Magic Diacritic Key" then?

Nice..., luckily only U has more than one.

**Reply by iandoug** [reply]
For comparison, German acrobatics.

The characters shown at the right border of a keytop are accessed by first pressing a dead key sequence of AltGr plus the × multiplication sign. This X-like symbol may be thought of as an "extra" dead key or "extra" accent type, used to access "miscellaneous" letters that do not have a specific accent type like diaeresis or circumflex. Symbols on the right border shown in green have both upper-case and lower-case forms; the corresponding capital letter is available by pressing the Shift key simultaneously with the symbol key. For instance, to type the lower-case æ ligature, hold the AltGr key and type ×, then release both keys and type the (unshifted) A key. To type the upper-case Æ ligature, hold the AltGr key and type ×, then release both 
keys, hold Shift and type the (shifted) A key. An active Caps Lock can be used instead of the Shift key to obtain the Æ ligature and similar letters. 

https://en.wikipedia.org/wiki/German_keyboard_layout

**Reply by iandoug** [reply]
★ U+2605
☆ U+2606
✦ U+2726
✧ U+2727
≛ U+225B

or something else?

**Reply by binarybottle** [reply]
@iandoug -- I was hoping to simply press a special function key, like a shift key but centrally placed, and the subsequent key or simultaneous key combination would map to a symbol of our choice. Is it possible to create this special function key?

**Reply by iandoug** [reply]
I suspect it is going to depend on 'drivers' to do it.

Take a look at Kmonad. https://github.com/kmonad/kmonad

Judging from the description of how the existing tilde char is used, it is certainly possible.

KLA has no concept of such things, I have started thinking how to add it .. but not figured it out yet.

**Reply by iandoug** [reply]
On the other hand, their separate dead keys must have a use, even if for non-Spanish, like correspondence with France needing ë.

**Reply by binarybottle** [reply]
I was hoping to get it to work with Keyman for super-simple installation as well.

**Reply by binarybottle** [reply]
@iandoug -- By "Magic Diacritic Key", do you mean the type of Shift-like "special function key" I'm referring to?

**Reply by iandoug** [reply]
Yes... where it triggers different diacritics depending on next letter. So you don't get ņ or ę but do get ç. Etc.

**Reply by Lobo-Feroz** [reply]
> What do you think about the following strategy, @NickG13, @Lobo-Feroz, and @sunaku?
> All diacriticals and some symbols are accessed by a special key (denoted by a star):

I like this model, but I'm not seeing how is the ★ key different to the ´ key which we already have and could do this (and does already with the vowels).

As per your comments it seems it's because this star would be used simultaneously, as another layer, instead of consecutively like the ´ key.

In any case you don't really need both ★ and ´. One of them could go, so that could save one key.

Single ´ could be written with (★ + space) so you don't really need that key. The grave accent or backtick ` is not really much used in spanish.

In this symbol layout proposal:

![engram-es-v1-normal](https://user-images.githubusercontent.com/10242383/127784971-f58af444-ca1b-4b05-a25a-45087ac41ac7.png)

That ´ or ★ key would be in the top row, center right column. Instead of the apostrophe, and switched sides with " because all vowels are in the left. I can see it can be very similar to your approach, it can be simultaneous instead of consecutive, and can have another symbol to label it, instead of the acute accent.

**Reply by Lobo-Feroz** [reply]
> so "A Magic Diacritic Key" then?

From what I gather from your comments, there's basically one hurdle here: decide if it's gonna be simultaneous (layered) or consecutive ("magic key").

Do you know if there is any reference anywhere comparing these two modes? Or alternatively, do people here have any preference?
 
> Nice..., luckily only U has more than one.

Absolutely! And luckily the current tilde key for qwerty-es already has this solved. The shift for ´ is ¨, so Ü is Shift+´, then U.

This would work only for the consecutive mode, since for simultaneous mode you couldn't differentiate ü from Ü, but then you could do ★ + : = Ü like Arno says or add AltGr, wich would be a weird combo of AltGr+★+Shift+u. Fingerbreaker!

Fortunately I think this case is quite rare. And let me explain why. ü is used to make sounding an otherwhise silent u in the syllabes **gue**, **gui**.

Let's take the word "pingüino" (penguin). It means that ü has sound. If it had no dieresis, "pinguino", the u would be silent.

Which means ü only happens _within_ the syllabes **gue**, **gui**, which means it won't happen at the beginning of a word. I.e. there are no words that start with Ü. Wich means the only way to have a capital Ü would be to have all characters in capitals, for which you set caps lock and that's it. Unless you are doing sOmE WeIrD ThInG WiTh cApItALS, in wich case you deserve the fingerbreaking.

Edit: after having a look at Ian's monogram frequencies in the new thread, this hypothesis seems correct, because capital Ü has a frequency of 2.0E-5. It's near the bottom of the list.

A different issue would be if you want a general solution for a magic diacritic key for different diacritics for different languages. Which might be a good idea if you plan to do at some point engram-de and others. That might be a little more complex.

**Reply by binarybottle** [reply]
Thank you for clarifying, @Lobo-Feroz!  

Re: "And luckily the current tilde key for qwerty-es already has this solved. The shift for ´ is ¨, so Ü is Shift+´, then U.":
I think this is an elegant solution. But how does one type an apostrophe (´)?

**Reply by Lobo-Feroz** [reply]
Well, since the apostrophe is not a very frequent character (as per Ian's frequency calculations), we could set it up in a mnemonic place like you do with %, & and the like.

An idea would be to add it to the quote key ", rationale:

- In spanish the apostrophe is akin to the single quote, and the quote signs are, in order of canonical priority (« ») (“ ”) (‘’). 
  - Tough as I've said before, many people just use mostly " as primary since qwerty-es lacks guillemets. But even considering this, the ' would still be secondary quotes to "  
- Visually, they look similar, like ' being the single version of "
- For strict spanish it doesn't matter, but sometimes to write something in english, since ' in english is very frequent, maybe it's not a good idea to send ' to fingerbreaking hell. Maybe a secondary placement is enough.

So maybe it could be the (tilde|star) version of ", or the AltGr version of "

Edit: btw, you wrote "apostrophe" (') but the symbol you wrote was "accute accent" (´). Did you mean this latter one? That one would be (tilde|star)+space, just like the current tilde key.

**Reply by binarybottle** [reply]
@Lobo-Feroz -- Sorry, I did mean the acute accent (´); I misread that as an apostrophe.  I have the apostrophe taken care of above: star + » = '  

What did you mean by: "[the acute accent (´)] would be (tilde|star)+space, just like the current tilde key."?

**Reply by iandoug** [reply]
Think we might as well pick a different symbol/name for this key to avoid confusion with asterisk * .

**Reply by binarybottle** [reply]
How about a "diacritikey"?

**Reply by iandoug** [reply]
Dire Criticky ? :-)

**Reply by Lobo-Feroz** [reply]
> What did you mean by: "[the acute accent (´)] would be (tilde|star)+space, just like the current tilde key."?

OK @binarybottle we are going around in circles here. Let me explain from the beginning how diacritics work in the current qwerty-es keyboard.

Here's the qwerty-es with the tilde key highlighted in green. Between Ñ and Enter:

![qwerty-es-tilde](https://user-images.githubusercontent.com/10242383/127929066-172b5fd6-d16a-439b-9698-ff4d165c6893.png)

And here's how it looks in an actual keyboard:

![tilde-key-200](https://user-images.githubusercontent.com/10242383/127930515-ead6006b-6e28-40ef-8131-a71dbbaa1ccd.jpg)

It works consecutively. You hit this key then the key you want tilded.

In base layer, the character is the accute accent ´ , wich for us spaniards is coloquially called "tilde" (even if tilde, strictly, is a name for all these diacritical marks, we all call this "tilde"). So in base layer you hit ´+ A = Á. So you can have ÁÉÍÓÚ.

In shifted layer, it is the Diaeresis ¨   . So Shift + ´ + A = Ä. So you can have ÄËÏÖÜ.

I think this is a nice working model. My proposal was to extend it so that ´+ N = Ñ, so that we free the Ñ key for something else. This is because Ñ has a very low frequency. Same idea for ´+ C = Ç. 2 keys freed.

If you want just the single ´ or ¨ symbols, you just hit that key, base or shifted, then space.

So I think the current workings of this tilde key are functional enough for having most of our diacritical marks. At least if it's used sequentially as is the case now with the current qwerty-es.

------

It might be more comfortable hitting the tilde key and the key to be tilded simultaneously instead of sequentially. It will be faster at the very least. It introduces a problem.

The problem is that when hitting those and shift at the same time, it's the same combo for a tilded capital vowel like Á (shifted a, tilded), than for ä (a, shifted tilded), than for Ä (shifted a, shifted tilded). So that would need a workaround. In spanish it's not very problematic since we only use ü and 99.9% of the time it's untilded, but if you want to extend engram to many other languages keeping the diacritics method, we might need an universal solution.

**Reply by binarybottle** [reply]
Thank you so much for the clarification! The picture helped! OK, I really like the idea of a sequential key taking care of all diacriticals.  I'm sold. Let's do it that way.

**Reply by iandoug** [reply]
> It works consecutively. You hit this key then the key you want tilded.
> 
> In base layer, the character is the accute accent ´ , wich for us spaniards is coloquially called "tilde" (even if tilde, strictly, is a name for all these diacritical marks, we all call this "tilde"). So in base layer you hit ´+ A = Á. So you can have ÁÉÍÓÚ.
> 
> In shifted layer, it is the Diaeresis ¨ . So Shift + ´ + A = Ä. So you can have ÄËÏÖÜ.
> 
> I think this is a nice working model. My proposal was to extend it so that ´+ N = Ñ, so that we free the Ñ key for something else. This is because Ñ has a very low frequency. Same idea for ´+ C = Ç. 2 keys freed.

I think part of the problem here was terminology.
That symbol is not tilde, it's a back-tick or grave accent. Tide is "~", and sits and looks like this on ANSI:

![2021-08-03_08-57](https://user-images.githubusercontent.com/12673886/127971524-e7662609-f3e6-4187-966d-a1612b8d463f.png)

So I thought previous references to "the tilde key" meant that key. The Germans also use that key for their magic diacritics.

I also understood from the original description that it was already capable of doing é, ü, ñ or ç depending on letter, not that this is what you would LIKE it to do.

Glad it's all cleared up now :-)

**Reply by Lobo-Feroz** [reply]
> I think part of the problem here was terminology.
> That symbol is not tilde, it's a back-tick or grave accent. Tide is "~", and sits and looks like this on ANSI:

Yes, my fault. In spanish our tilde is your english accute accent (not grave) ([wiki-es](https://es.wikipedia.org/wiki/Acento_gr%C3%A1fico)). And your tilde is our "virgulilla" ([wiki-es](https://es.wikipedia.org/wiki/Virgulilla)). But since we're communicating in english I should have used english terminology, sorry. I just didn't notice. Accute accent it is.

> I also understood from the original description that it was already capable of doing é, ü, ñ or ç depending on letter, not that this is what you would LIKE it to do.

Yeah it was an idea I wrote in the initial thead in the engram-en repo. It's buried somewhere in the "Original discussion" thread here where Arno copied all the original messages.

With the accute accent diacritical key you can make almost all the spanish diacritics: áéíóú ÁÉÍÓÚ üÜ
The only ones missings are ñÑ çÇ. Which currently have their own keys. 
But they are very very low frequency, that's why propose doing them with the accute accent key as well. (It's just a proposal btw, of course there can be other ways to write these).

According to your current monogram frequencies txt:
```
ñ	4184439	0.15183
ç	23417	0.00085
Ñ	5720	0.00021
Ç	283	1.0E-5
```

They are way way lower than the frequencies of n and c:
```
n	154603794	5.60975
c	93357395	3.38744
C	5967779         0.21654
N	2282144	        0.08281
```
------

Now on the **grave accent** or **backtick**. 

![gave-accent-key](https://user-images.githubusercontent.com/10242383/127978873-3c7ca4cf-41f3-4842-81ff-70ef116a99b6.png)

This is a different key. It sits upwards of the accute accent key and its shifted version is the circumflex ^ 

You can use them to write: àèìòù ÀÈÌÒÙ âêîôû ÂÊÎÔÛ

Here's the deal: none of these, none, are valid characters in spanish. But I'm sure they are valid characters in other languages. For instance, French uses all of these: https://jakubmarian.com/french-e-e-e-e-e-whats-the-difference/

So maybe @binarybottle is interested in designing right now a comprehensive solution so that it might be valid for several languages, maybe only switching keys around because of frequencies.

> Glad it's all cleared up now :-)

:) - so, consecutive typing with the accute accent key it is then.

**Reply by binarybottle** [reply]
@Lobo-Feroz -- Thank you for the suggestion of a comprehensive solution. I will look into this!

**Reply by NickG13** [reply]
grave accent is very used in Catalan. Some of the spanish typist are bilingual, so that's why ` exists. There's also some special characters in some dialects, like asturian which has some kind of dot in letters. I've never seen asturian text tho.

**Comment by iandoug** [comment]
>     9\. ¨ - Diaeresis. Makes a silent u vowel have sound. https://en.wikipedia.org/wiki/Diaeresis_(diacritic)

Is this only a dead key? I found a few in the Leipzig corpus (3693) and 1 in the Gutenberg, but it may be the result of mangled Unicode...

Currently populating the default Qwerty-es with all the new symbols as a reference.

Thanks, Ian

**Reply by Lobo-Feroz** [reply]
I'm not sure about what you mean with "only a dead key".

Can ¨ appear on its own? Only as the symbol itself.

Is ¨ used in the language anywhere, without a vowel, to form words or sentences? No, never. 

Let me know if this is what you were asking, please.

**Reply by iandoug** [reply]
Yes ... it should not exist in normal text then.
And is only used in Spanish for ü.
So if we do the (rare) ü in a different way (eg AltGr), then there is no need for it on the keyboard?

I use " to put diaeresis on letters when needed. ë ö etc

**Reply by Lobo-Feroz** [reply]
Yeah. As you know in the current kb is a dead key, shifted accute accent.

If we move the accute accent to the middle columns, for instance where the apostrophe is, the shifted version would conflict with the parenthesis (as per [engram-es](https://raw.githubusercontent.com/binarybottle/engram/master/assets/engram-800px.png)).

I don't know if engram-es has an AltGr layer defined for that location, it might be a good solution. AltGr + ´ would become the ¨ dead key, and from that it could be normally used, for all its current combinations.

I'm looking into keyboard layout setup right now. I don't know if implementing that would be possible. Let me try and see.

**Reply by Lobo-Feroz** [reply]
Yes, we can set up diaeresis as the AltGr version of the accute accent, and make it a dead key:

![dead-key-diaeresis](https://user-images.githubusercontent.com/10242383/129412344-2942b54c-4618-4453-b2ec-98ceb0373ed5.png)
